dali/kernels/common/pad.h

```diff
@@ -0,0 +1,145 @@
+        if (dim > biggest_dim)
+          biggest_dim = dim;
+      }
+      auto ts = in.tensor_shape(0);
+      ts[axis] = biggest_dim;
+      std::vector<TensorShape<DynamicDimensions>> tl_shape;
+      for (int i = 0; i < num_samples; ++i) {
+        tl_shape.push_back(ts);
+      }
+      req.output_shapes = {TensorListShape<DynamicDimensions>(tl_shape)};
+      block_count_ = std::ceil((volume(ts) * num_samples) / static_cast<float>(kBlockSize));
+
+      ScratchpadEstimator se;
+      se.add<detail::SampleDesc<Dims, Type>>(AllocType::Host, num_samples);
+      se.add<detail::SampleDesc<Dims, Type>>(AllocType::GPU, num_samples);
+
+      se.add<detail::BlockDesc>(AllocType::Host, block_count_);
+      se.add<detail::BlockDesc>(AllocType::GPU, block_count_);
+      req.scratch_sizes = se.sizes;
+    }
+    return req;
+  }
+
+  DLL_PUBLIC void Run(KernelContext &context, OutListGPU<Type, Dims> &out,
+                      const InListGPU<Type, Dims> &in, Type padding_val) {
+    auto num_samples = static_cast<size_t>(in.num_samples());
+
+    detail::SampleDesc<Dims, Type>* sample_descs_cpu =
+      context.scratchpad->Allocate<detail::SampleDesc<Dims, Type>>(AllocType::Host,
+                                                                         num_samples);
+    detail::BlockDesc *block_descs_cpu =
+      context.scratchpad->Allocate<detail::BlockDesc>(AllocType::Host, block_count_);
+
+    std::vector<size_t> sample_sizes(in.size());
+    for (int i = 0; i < in.size(); i++) {
+      const auto in_shape = in.tensor_shape(i);
+      auto& out_shape = out.tensor_shape(i);
+
+      auto &sample_desc = sample_descs_cpu[i];
+      sample_desc.in_strides = GetStrides<Dims>(in_shape);
+      sample_desc.out_strides = GetStrides<Dims>(output_shape);
+      // Weird implem: we rely only on strides but we need out_shape != padded_out_shape
+      sample_desc.out_shape = in_shape;
+      sample_desc.padded_out_shape = out_shape;
+      sample_desc.padding_val = padding_val;
+      sample_desc.in = in.tensor_data(i);
+      sample_desc.out = out.tensor_data(i);
+      sample_sizes[i] = volume(out_shape);
+    }
+
+    // Calculate super-block offset and size
+    size_t block_idx = 0;
+    for (int i = 0; i < static_cast<int>(num_samples); i++) {
+      size_t offset = 0;
+      size_t remaining = sample_sizes[i];
+      while (remaining > 0) {
+        size_t size = remaining < kBlockSize ? remaining : kBlockSize;
+        block_descs_cpu[block_idx++] = {i, offset, size};
+        remaining -= size;
+        offset += size;
+      }
+    }
+
+    // Transfer to GPU
+    detail::SampleDesc<Dims, Type> *sample_descs =
+      context.scratchpad->Allocate<detail::SampleDesc<Dims, Type>>(
+        AllocType::GPU, num_samples);
+    detail::BlockDesc *block_descs =
+      context.scratchpad->Allocate<detail::BlockDesc>(
+        AllocType::GPU, block_count_);
+
+    // Memory is allocated contiguously, so we launch only one cudaMemcpyAsync
```

Comment:
Do we have such guarantee? Or it is how it happens now.

Reply:
I copy pasted this line from Slice op
AFAIK this is guaranteed by the ScratchPad implementation